Sources/Profile-View.php

```diff
@@ -217,221 +217,353 @@ function summary($memID)
+
+			// Keep track of this so we can use it below.
+			if (isset($data['board_name']))
+				$alert['extra']['board_name'] = $data['board_name'];
+
+			unset($alert['content_data']);
 		}
-		else
-			$alert_topic_msg = $txt['topic_na'];
 
-		// Check that there's a valid board
-		if (isset($alert['extra']['board']) && !empty($board_msgs[$alert['extra']['board']]))
-			$alert_board_msg = $board_msgs[$alert['extra']['board']][$alert['show_links'] ? 'link' : 'text'];
-		elseif (isset($alert['extra']['board']))
+		// Do we want to link to the topic in general or the new messages specifically?
+		if (isset($possible_topics[$id_alert]))
+			$alert['extra']['topic_suffix'] = 'new#new';
+		elseif (isset($alert['extra']['topic']))
+			$alert['extra']['topic_suffix'] = '0';
+
+		// Make sure profile alerts have what they need.
+		if (in_array($id_alert, $profile_alerts))
 		{
-			unset($alerts[$id_alert]);
-			continue;
+			if (empty($alert['extra']['user_id']))
+				$alert['extra']['user_id'] = $alert['content_id'];
+
+			if (isset($user_profile[$alert['extra']['user_id']]))
+				$alert['extra']['user_name'] = $user_profile[$alert['extra']['user_id']]['real_name'];
 		}
-		else
-			$alert_board_msg = $txt['board_na'];
 
-		// Set the texts
-		$alerts[$id_alert]['extra']['msg_msg'] = $alert_msg_msg;
-		$alerts[$id_alert]['extra']['topic_msg'] = $alert_topic_msg;
-		$alerts[$id_alert]['extra']['board_msg'] = $alert_board_msg;
+		// Next, build the message strings.
+		foreach ($formats as $msg_type => $format_info)
+		{
+			// Get the values to use in the formatted string, in the right order.
+			$msg_values = array_replace(
+				array_fill_keys($format_info['required'], ''),
+				array_intersect_key($alert['extra'], array_flip($format_info['required']))
+			);
 
-		if ($alert['content_type'] == 'profile')
-			$alerts[$id_alert]['extra']['profile_msg'] = $alert['show_links'] ? '<a href="' . $scripturl . '?action=profile;u=' . $alerts[$id_alert]['content_id'] . '">' . $alerts[$id_alert]['extra']['user_name'] . '</a>' : '<strong>' . $alerts[$id_alert]['extra']['user_name'] . '</strong>';
+			// Assuming all required values are present, build the message.
+			if (!in_array('', $msg_values))
+				$alert['extra'][$msg_type] = vsprintf($formats[$msg_type][$alert['show_links'] ? 'link' : 'text'], $msg_values);
 
-		if (!empty($memberContext[$alert['sender_id']]))
-			$alerts[$id_alert]['sender'] = &$memberContext[$alert['sender_id']];
+			elseif (in_array($msg_type, array('msg_msg', 'topic_msg', 'board_msg')))
+				$alert['extra'][$msg_type] = $txt[$msg_type == 'board_msg' ? 'board_na' : 'topic_na'];
+			else
+				$alert['extra'][$msg_type] = '(' . $txt['not_applicable'] . ')';
+		}
 
+		// If we loaded the sender's profile, we may as well use it.
+		$sender_id = !empty($alert['sender_id']) ? $alert['sender_id'] : 0;
+		if (isset($user_profile[$sender_id]))
+			$alert['sender_name'] = $user_profile[$sender_id]['real_name'];
```

Comment:
don't forget we need the sender's avatar though, I think we can get that after loading the member's context

Reply:
Bleh. You're right. I guess I'll put it back, then.

Reply:
I've restored the code to load the `$memberContext` info (including the avatar) for the alert senders, but before I commit it I'd appreciate interaction on my question below.